runtime/notification.go

```diff
@@ -0,0 +1,71 @@
+package runtime
+
+import (
+	"fmt"
+
+	"github.com/mitchellh/mapstructure"
+)
+
+type Notification struct {
+	GroupID   string      `json:"group_id"`
+	EventType string      `json:"event_type"`
+	Event     interface{} `json:"event"`
+}
+
+type StageEvent struct {
+	Name   string `json:"name"`
+	Action string `json:"action"` // entry or exit
+}
+
+type OutcomeEvent struct {
+	Outcome string `json:"outcome"` // ok
+}
+
+func NewStageNotification(groupId, name, action string) *Notification {
+	return &Notification{
+		GroupID:   groupId,
+		EventType: "stage",
+		Event: &StageEvent{
+			Name:   name,
+			Action: action,
+		},
+	}
+}
+
+func NewOutcomeOKNotification(groupId string) *Notification {
+	return &Notification{
+		GroupID:   groupId,
+		EventType: "outcome",
+		Event: &OutcomeEvent{
+			Outcome: "ok",
+		},
+	}
+}
```

Comment:
This pattern seems to suggest we plan to create a constructor per each possible event type... I think that leads to unnecessary proliferation of bespoke constructors. I'd rather have a single, flat event struct type that we can easily populate inline, or with a single constructor that accepts options. Inspiration: zap.

Reply:
`single, flat event struct type` -> I considered that (as this is what the `Event` is doing), but I find it more confusing. You never know which fields will be filed by which type of event, whereas here we have it type-safe, once we know what the `EventType` is.

```
type Event struct {
	Type       EventType    `json:"type"`
	Outcome    EventOutcome `json:"outcome,omitempty"`
	Error      string       `json:"error,omitempty"`
	Stacktrace string       `json:"stacktrace,omitempty"`
	Message    string       `json:"message,omitempty"`
```

For example some events don't have an Outcome, some events don't have a Message or a Stacktrace, etc. - it is not really clear which fields we can rely on being present (i.e. required), and which are optional.

In any case this is up for discussion on how exactly we want to propagate that information.

Reply:
Right, but from the perspective of the consumer of this payload, a polymorphic type and a flattened type are equivalent in terms of the information the consumer needs to have in order to process the payload. In both cases, a decision needs to be made.

- polymorphic: look at a "event type" field, and deserialise the generic payload as one type or another depending on the value of the former.
- flattened: look at the "event type field", and process one field or another based on it.

Both approaches are equivalent in terms of the burden that's placed on the consumer.

In these cases, I usually look at how much variance we'll need to support. In this case, there are _really_ only two or three kinds/categories of events, so the polymorphic approach is too heavy-weight IMO. That's why I went with the flat struct.

In cases where the event types are unbounded, open-ended, or some form of decoupling is needed where we want to make it possible for the consumer to receive event types they don't know how to deserialise (e.g. to future-proof p2p protocols), I would absolutely choose a polymorphic approach. A recent example of that is how the Lotus journal is implemented: https://github.com/filecoin-project/lotus/blob/master/journal/types.go#L88

But I don't think we need that level of genericness here.

Reply:
> Right, but from the perspective of the consumer of this payload, a polymorphic type and a flattened type are equivalent in terms of the information the consumer needs to have in order to process the payload. In both cases, a decision needs to be made.

Totally agree.

> Both approaches are equivalent in terms of the burden that's placed on the consumer.

Also agree in terms of consumer/program. I'd argue as a human/developer I find it hard to process the types of Events we have with a flat structure. Every time I look at the Event and consider the variations we have, I am confused. In my opinion (and I realise this most probably is personal preference) it is much simpler to have a struct for each. So the type-safety help readability.

---

I agree that now we have two/three kinds/categories of events, but flattening them makes is harder to reason about when reading the code, specially when they are rather different from each other.